# Here is the notebook for all of the scraping efforts

#### #note: All the files used to be on this folder. Since the efforts are finished, they no longer refer to the correct locations.

### <marquee><u>*Goals*</u></marquee>

My overall goal was to get the career stats of every player, and their stats for the last 3 seasons. I only wanted active players and non rookies for simplicity. I achieved this by scraping the list of teams, collecting all the active players on each of teams, then going to the individual players pages and scraping from there. I placed the career and the last three years stats on two different data frame

In [1]:
import pandas as pd
import requests #gets data from websites
from bs4 import BeautifulSoup
import time

In [2]:
pd.set_option("display.max_columns",40)

### Here is the process of scraping the list of teams

In [3]:
url='https://www.basketball-reference.com/teams/'

In [2]:
res = requests.get(url)
team_page=BeautifulSoup(res.content,'lxml')

In [5]:
team_page

<!DOCTYPE html>
<html class="no-js" data-root="/home/bbr/build" data-version="klecko-" itemscope="" itemtype="https://schema.org/WebSite" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="ie=edge" http-equiv="x-ua-compatible"/>
<meta content="width=device-width, initial-scale=1.0, maximum-scale=2.0" name="viewport"/>
<link href="https://d2p3bygnnzw9w3.cloudfront.net/req/201907110" rel="dns-prefetch"/>
<!-- no:cookie fast load the css.           -->
<link crossorigin="" href="https://d2p3bygnnzw9w3.cloudfront.net" rel="preconnect"/>
<link crossorigin="" href="https://d2cwpp38twqe55.cloudfront.net" rel="preconnect"/>
<style>html,body{margin:0;padding:0;font:14px/1.25 "Helvetica Neue",helvetica,arial,sans-serif;color:#000}body{position:relative;background:#c9cbcd;z-index:0;-webkit-text-size-adjust:none;-moz-text-size-adjust:none;-ms-text-size-adjust:none}a img{border:0}ul,li,ol{margin:0;padding:0;list-style-type:none}table th,table td{border:0}iframe{max-width:100%}code{background:

In [15]:
team_page_list= []
for team in team_page.find_all('tr',{'class':"full_table"})[0:30]:
    instance={}
    instance['team_name']=team.find('a').text
    instance['url']=team.find('th').find('a').attrs['href'].replace('/teams/',"")
    team_page_list.append(instance)
    
teams_df=pd.DataFrame(team_page_list)

In [16]:
teams_df

team_name   url
0            Atlanta Hawks  ATL/
1           Boston Celtics  BOS/
2            Brooklyn Nets  NJN/
3        Charlotte Hornets  CHA/
4            Chicago Bulls  CHI/
5      Cleveland Cavaliers  CLE/
6         Dallas Mavericks  DAL/
7           Denver Nuggets  DEN/
8          Detroit Pistons  DET/
9    Golden State Warriors  GSW/
10         Houston Rockets  HOU/
11          Indiana Pacers  IND/
12    Los Angeles Clippers  LAC/
13      Los Angeles Lakers  LAL/
14       Memphis Grizzlies  MEM/
15              Miami Heat  MIA/
16         Milwaukee Bucks  MIL/
17  Minnesota Timberwolves  MIN/
18    New Orleans Pelicans  NOH/
19         New York Knicks  NYK/
20   Oklahoma City Thunder  OKC/
21           Orlando Magic  ORL/
22      Philadelphia 76ers  PHI/
23            Phoenix Suns  PHO/
24  Portland Trail Blazers  POR/
25        Sacramento Kings  SAC/
26       San Antonio Spurs  SAS/
27         Toronto Raptors  TOR/
28               Utah Jazz  UTA/
29      Washington Wizards  WAS/

In [61]:
teams_df.to_csv('teams_and_slugs_df.csv',index=False)

In [6]:
teams_df = pd.read_csv('teams_and_slugs_df.csv')

In [7]:
teams_df

team_name   url
0            Atlanta Hawks  ATL/
1           Boston Celtics  BOS/
2            Brooklyn Nets  NJN/
3        Charlotte Hornets  CHA/
4            Chicago Bulls  CHI/
5      Cleveland Cavaliers  CLE/
6         Dallas Mavericks  DAL/
7           Denver Nuggets  DEN/
8          Detroit Pistons  DET/
9    Golden State Warriors  GSW/
10         Houston Rockets  HOU/
11          Indiana Pacers  IND/
12    Los Angeles Clippers  LAC/
13      Los Angeles Lakers  LAL/
14       Memphis Grizzlies  MEM/
15              Miami Heat  MIA/
16         Milwaukee Bucks  MIL/
17  Minnesota Timberwolves  MIN/
18    New Orleans Pelicans  NOH/
19         New York Knicks  NYK/
20   Oklahoma City Thunder  OKC/
21           Orlando Magic  ORL/
22      Philadelphia 76ers  PHI/
23            Phoenix Suns  PHO/
24  Portland Trail Blazers  POR/
25        Sacramento Kings  SAC/
26       San Antonio Spurs  SAS/
27         Toronto Raptors  TOR/
28               Utah Jazz  UTA/
29      Washington Wizards  WAS/

In [8]:
team_page_list=teams_df.to_dict('records')

In [9]:
team_page_list

[{'team_name': 'Atlanta Hawks', 'url': 'ATL/'},
 {'team_name': 'Boston Celtics', 'url': 'BOS/'},
 {'team_name': 'Brooklyn Nets', 'url': 'NJN/'},
 {'team_name': 'Charlotte Hornets', 'url': 'CHA/'},
 {'team_name': 'Chicago Bulls', 'url': 'CHI/'},
 {'team_name': 'Cleveland Cavaliers', 'url': 'CLE/'},
 {'team_name': 'Dallas Mavericks', 'url': 'DAL/'},
 {'team_name': 'Denver Nuggets', 'url': 'DEN/'},
 {'team_name': 'Detroit Pistons', 'url': 'DET/'},
 {'team_name': 'Golden State Warriors', 'url': 'GSW/'},
 {'team_name': 'Houston Rockets', 'url': 'HOU/'},
 {'team_name': 'Indiana Pacers', 'url': 'IND/'},
 {'team_name': 'Los Angeles Clippers', 'url': 'LAC/'},
 {'team_name': 'Los Angeles Lakers', 'url': 'LAL/'},
 {'team_name': 'Memphis Grizzlies', 'url': 'MEM/'},
 {'team_name': 'Miami Heat', 'url': 'MIA/'},
 {'team_name': 'Milwaukee Bucks', 'url': 'MIL/'},
 {'team_name': 'Minnesota Timberwolves', 'url': 'MIN/'},
 {'team_name': 'New Orleans Pelicans', 'url': 'NOH/'},
 {'team_name': 'New York Knic

In [14]:
team_page_list[2]['team_name']

'Brooklyn Nets'

In [17]:
team_page_list[:5]

[{'team_name': 'Atlanta Hawks', 'url': 'ATL/'},
 {'team_name': 'Boston Celtics', 'url': 'BOS/'},
 {'team_name': 'Brooklyn Nets', 'url': 'NJN/'},
 {'team_name': 'Charlotte Hornets', 'url': 'CHA/'},
 {'team_name': 'Chicago Bulls', 'url': 'CHI/'}]

In [18]:
for team in team_page_list:
    print (team['url'])

ATL/
BOS/
NJN/
CHA/
CHI/
CLE/
DAL/
DEN/
DET/
GSW/
HOU/
IND/
LAC/
LAL/
MEM/
MIA/
MIL/
MIN/
NOH/
NYK/
OKC/
ORL/
PHI/
PHO/
POR/
SAC/
SAS/
TOR/
UTA/
WAS/


##### I used try/except since some team urls lead to errors. This is due to some teams rebranding. 
##### ie: The Brooklyn Nets used to be the New Jersey Nets

In [30]:
team_rosters=[]
year='2020.html'
for team in team_page_list[0:]:
    try:
        res = requests.get(url+team['url']+'2020.html')

        soup = BeautifulSoup(res.content, 'lxml')
        table = soup.find('table',{'id':'roster'}).find('tbody')
        for row in table.find_all('tr')[0:]:        
            player={}        
            player['name']=row.find('td').find('a').text
            player['pos'] =row.find_all('td')[1].text
#            print(row.find('td').find('a').text)
            player['team']=team['team_name']
            player['slug']=row.find('td').find('a').attrs['href']
            player['exp'] =row.find_all('td')[6].text
            team_rosters.append(player)
            time.sleep(3)
    except:
        #used to identify which teams lead to errors
        print("blame the"+team['team_name'])

players_df = pd.DataFrame(team_rosters)

blame theBrooklyn Nets
blame theCharlotte Hornets
blame theNew Orleans Pelicans


In [31]:
players_df

exp                     name  pos                       slug  \
0     8         Chandler Parsons    F  /players/p/parsoch01.html   
1     9              Evan Turner    G  /players/t/turneev01.html   
2     6             Allen Crabbe  G-F  /players/c/crabbal01.html   
3     3          DeAndre' Bembry    F  /players/b/bembrde01.html   
4     3             Damian Jones  F-C  /players/j/jonesda03.html   
5     2             John Collins    F  /players/c/collijo01.html   
6     1               Trae Young    G  /players/y/youngtr01.html   
7     1            Kevin Huerter    G  /players/h/huertke01.html   
8     6                 Alex Len    C    /players/l/lenal01.html   
9     R          De'Andre Hunter    F  /players/h/huntede01.html   
10    R              Cam Reddish    F  /players/r/reddica01.html   
11    R           Bruno Fernando    F  /players/f/fernabr01.html   
12    5            Jabari Parker    F  /players/p/parkeja01.html   
13    3             Jaylen Brown  F-G  /players/b/brownja02.html   
14    2             Jayson Tatum    F  /players/t/tatumja01.html   
15    9           Gordon Hayward  F-G  /players/h/haywago01.html   
16    2             Semi Ojeleye    F  /players/o/ojelese01.html   
17    1          Robert Williams    F  /players/w/williro04.html   
18    5             Marcus Smart    G  /players/s/smartma01.html   
19    R           Romeo Langford    G  /players/l/langfro01.html   
20    R           Grant Williams    F  /players/w/willigr01.html   
21    8             Kemba Walker    G  /players/w/walkeke02.html   
22    R           Carsen Edwards    G  /players/e/edwarca01.html   
23    2             Daniel Theis    C  /players/t/theisda01.html   
24    8              Enes Kanter    C  /players/k/kanteen01.html   
25    1           Brad Wanamaker    G  /players/w/wanambr01.html   
26    2         Denzel Valentine    G  /players/v/valende01.html   
27    3                Kris Dunn    G   /players/d/dunnkr01.html   
28    2          Lauri Markkanen  F-C  /players/m/markkla01.html   
29    4        Cristiano Felício  F-C  /players/f/feliccr01.html   
..   ..                      ...  ...                        ...   
355   4          Stanley Johnson    F  /players/j/johnsst04.html   
356   4  Rondae Hollis-Jefferson    F  /players/h/holliro01.html   
357   6              Rudy Gobert    C  /players/g/goberru01.html   
358  12              Mike Conley    G  /players/c/conlemi01.html   
359   2         Donovan Mitchell    G  /players/m/mitchdo01.html   
360   2             Tony Bradley  F-C  /players/b/bradlto01.html   
361   2            Royce O'Neale    F  /players/o/onealro01.html   
362   5               Joe Ingles  F-G  /players/i/inglejo01.html   
363   4               Dante Exum    G   /players/e/exumda01.html   
364   3            Georges Niang    F  /players/n/niangge01.html   
365   R                 Miye Oni    G    /players/o/onimi01.html   
366   5         Bojan Bogdanović  G-F  /players/b/bogdabo02.html   
367   R         Jarrell Brantley    F  /players/b/brantja01.html   
368   R    Justin Wright-Foreman    G  /players/w/wrighju02.html   
369  11              Ian Mahinmi    C  /players/m/mahinia01.html   
370   7             Bradley Beal    G   /players/b/bealbr01.html   
371  14               C.J. Miles  G-F  /players/m/milescj01.html   
372   9                John Wall    G   /players/w/walljo01.html   
373   1            Moritz Wagner  F-C  /players/w/wagnemo01.html   
374   1           Troy Brown Jr.  G-F  /players/b/browntr01.html   
375   1              Isaac Bonga    G  /players/b/bongais01.html   
376   3            Dāvis Bertāns    F  /players/b/bertada01.html   
377   1           Jemerrio Jones    F  /players/j/jonesje01.html   
378   3             Jordan McRae    G  /players/m/mcraejo01.html   
379   R            Tarik Phillip    G  /players/p/phillta01.html   
380   R            Rui Hachimura    F  /players/h/hachiru01.html   
381   R        Admiral Schofield  G-F  /players/s/schofad01.

In [ ]:
#bad teams Brooklyn Nets, Charlotte Hornets, New orleans Pelicans

In [26]:
players_df['team'].value_counts()

Memphis Grizzlies         18
Sacramento Kings          18
San Antonio Spurs         17
Washington Wizards        16
Detroit Pistons           16
Chicago Bulls             16
New York Knicks           16
Oklahoma City Thunder     15
Philadelphia 76ers        15
Los Angeles Lakers        14
Miami Heat                14
Milwaukee Bucks           14
Dallas Mavericks          14
Portland Trail Blazers    14
Denver Nuggets            14
Toronto Raptors           14
Golden State Warriors     13
Phoenix Suns              13
Cleveland Cavaliers       13
Houston Rockets           13
Minnesota Timberwolves    13
Orlando Magic             13
Atlanta Hawks             13
Boston Celtics            13
Los Angeles Clippers      12
Indiana Pacers            12
Utah Jazz                 12
Name: team, dtype: int64

In [32]:
players_df.to_csv("players_in_teams.csv",index=False)

In [30]:
players_df = pd.read_csv("players_in_teams.csv")

In [34]:
players_df

exp                     name  pos                       slug  \
0     8         Chandler Parsons    F  /players/p/parsoch01.html   
1     9              Evan Turner    G  /players/t/turneev01.html   
2     6             Allen Crabbe  G-F  /players/c/crabbal01.html   
3     3          DeAndre' Bembry    F  /players/b/bembrde01.html   
4     3             Damian Jones  F-C  /players/j/jonesda03.html   
5     2             John Collins    F  /players/c/collijo01.html   
6     1               Trae Young    G  /players/y/youngtr01.html   
7     1            Kevin Huerter    G  /players/h/huertke01.html   
8     6                 Alex Len    C    /players/l/lenal01.html   
9     R          De'Andre Hunter    F  /players/h/huntede01.html   
10    R              Cam Reddish    F  /players/r/reddica01.html   
11    R           Bruno Fernando    F  /players/f/fernabr01.html   
12    5            Jabari Parker    F  /players/p/parkeja01.html   
13    3             Jaylen Brown  F-G  /players/b/brownja02.html   
14    2             Jayson Tatum    F  /players/t/tatumja01.html   
15    9           Gordon Hayward  F-G  /players/h/haywago01.html   
16    2             Semi Ojeleye    F  /players/o/ojelese01.html   
17    1          Robert Williams    F  /players/w/williro04.html   
18    5             Marcus Smart    G  /players/s/smartma01.html   
19    R           Romeo Langford    G  /players/l/langfro01.html   
20    R           Grant Williams    F  /players/w/willigr01.html   
21    8             Kemba Walker    G  /players/w/walkeke02.html   
22    R           Carsen Edwards    G  /players/e/edwarca01.html   
23    2             Daniel Theis    C  /players/t/theisda01.html   
24    8              Enes Kanter    C  /players/k/kanteen01.html   
25    1           Brad Wanamaker    G  /players/w/wanambr01.html   
26    2         Denzel Valentine    G  /players/v/valende01.html   
27    3                Kris Dunn    G   /players/d/dunnkr01.html   
28    2          Lauri Markkanen  F-C  /players/m/markkla01.html   
29    4        Cristiano Felício  F-C  /players/f/feliccr01.html   
..   ..                      ...  ...                        ...   
355   4          Stanley Johnson    F  /players/j/johnsst04.html   
356   4  Rondae Hollis-Jefferson    F  /players/h/holliro01.html   
357   6              Rudy Gobert    C  /players/g/goberru01.html   
358  12              Mike Conley    G  /players/c/conlemi01.html   
359   2         Donovan Mitchell    G  /players/m/mitchdo01.html   
360   2             Tony Bradley  F-C  /players/b/bradlto01.html   
361   2            Royce O'Neale    F  /players/o/onealro01.html   
362   5               Joe Ingles  F-G  /players/i/inglejo01.html   
363   4               Dante Exum    G   /players/e/exumda01.html   
364   3            Georges Niang    F  /players/n/niangge01.html   
365   R                 Miye Oni    G    /players/o/onimi01.html   
366   5         Bojan Bogdanović  G-F  /players/b/bogdabo02.html   
367   R         Jarrell Brantley    F  /players/b/brantja01.html   
368   R    Justin Wright-Foreman    G  /players/w/wrighju02.html   
369  11              Ian Mahinmi    C  /players/m/mahinia01.html   
370   7             Bradley Beal    G   /players/b/bealbr01.html   
371  14               C.J. Miles  G-F  /players/m/milescj01.html   
372   9                John Wall    G   /players/w/walljo01.html   
373   1            Moritz Wagner  F-C  /players/w/wagnemo01.html   
374   1           Troy Brown Jr.  G-F  /players/b/browntr01.html   
375   1              Isaac Bonga    G  /players/b/bongais01.html   
376   3            Dāvis Bertāns    F  /players/b/bertada01.html   
377   1           Jemerrio Jones    F  /players/j/jonesje01.html   
378   3             Jordan McRae    G  /players/m/mcraejo01.html   
379   R            Tarik Phillip    G  /players/p/phillta01.html   
380   R            Rui Hachimura    F  /players/h/hachiru01.html   
381   R        Admiral Schofield  G-F  /players/s/schofad01.

##### Scraping from the teams that had errors

In [19]:
nets_rosters=[]
year='2020.html'

res = requests.get('https://www.basketball-reference.com/teams/BRK/2020.html')

soup = BeautifulSoup(res.content, 'lxml')
table = soup.find('table',{'id':'roster'}).find('tbody')
for row in table.find_all('tr')[0:]:        
    player={}        
    player['name']=row.find('td').find('a').text
    player['pos'] =row.find_all('td')[1].text
    player['team']=team_page_list[2]['team_name']
    player['slug']=row.find('td').find('a').attrs['href']
    player['exp'] =row.find_all('td')[6].text
    nets_rosters.append(player)
    time.sleep(3)


nets_players_df = pd.DataFrame(nets_rosters)

In [20]:
nets_players_df.head()

exp               name  pos                       slug           team
0    3     Taurean Prince    F  /players/p/princta02.html  Brooklyn Nets
1    3       Caris LeVert  G-F  /players/l/leverca01.html  Brooklyn Nets
2    2      Jarrett Allen  C-F  /players/a/allenja01.html  Brooklyn Nets
3    1        Džanan Musa  G-F   /players/m/musadz01.html  Brooklyn Nets
4    1     Rodions Kurucs    F  /players/k/kurucro01.html  Brooklyn Nets
5    5         Joe Harris    G  /players/h/harrijo01.html  Brooklyn Nets
6    5  Spencer Dinwiddie    G  /players/d/dinwisp01.html  Brooklyn Nets
7    R    Nicolas Claxton    F  /players/c/claxtni01.html  Brooklyn Nets
8   12       Kevin Durant  F-G  /players/d/duranke01.html  Brooklyn Nets
9   11     DeAndre Jordan    C  /players/j/jordade01.html  Brooklyn Nets
10   8       Kyrie Irving    G  /players/i/irvinky01.html  Brooklyn Nets
11  11    Wilson Chandler    F  /players/c/chandwi01.html  Brooklyn Nets
12   9     Garrett Temple    G  /players/t/templga01.html  Brooklyn Nets
13   1        Theo Pinson    G  /players/p/pinsoth01.html  Brooklyn Nets
14   3        David Nwaba    G  /players/n/nwabada01.html  Brooklyn Nets
15   3     Henry Ellenson    F  /players/e/ellenhe01.html  Brooklyn Nets

In [21]:
team_page_list[3]['team_name']

'Charlotte Hornets'

In [22]:
hornets_rosters=[]
year='2020.html'

res = requests.get('https://www.basketball-reference.com/teams/CHO/2020.html')

soup = BeautifulSoup(res.content, 'lxml')
table = soup.find('table',{'id':'roster'}).find('tbody')
for row in table.find_all('tr')[0:]:        
    player={}        
    player['name']=row.find('td').find('a').text
    player['pos'] =row.find_all('td')[1].text
    player['team']=team_page_list[3]['team_name']
    player['slug']=row.find('td').find('a').attrs['href']
    player['exp'] =row.find_all('td')[6].text
    hornets_rosters.append(player)
    time.sleep(3)


hornets_players_df = pd.DataFrame(hornets_rosters)

In [24]:
hornets_players_df = pd.DataFrame(hornets_rosters)

In [25]:
hornets_players_df

exp                    name  pos                       slug  \
0    7  Michael Kidd-Gilchrist    F  /players/k/kiddgmi01.html   
1    6             Cody Zeller  C-F  /players/z/zelleco01.html   
2   11           Nicolas Batum  F-G  /players/b/batumni01.html   
3    8         Bismack Biyombo  C-F  /players/b/biyombi01.html   
4   14         Marvin Williams    F  /players/w/willima02.html   
5    3       Willy Hernangómez    C  /players/h/hernawi01.html   
6    2              Malik Monk    G   /players/m/monkma01.html   
7    2            Dwayne Bacon    G  /players/b/bacondw01.html   
8    1           Miles Bridges  G-F  /players/b/bridgmi02.html   
9    1         Devonte' Graham    G  /players/g/grahade01.html   
10   R           PJ Washington    F  /players/w/washipj01.html   
11   4            Terry Rozier    G  /players/r/roziete01.html   

                 team  
0   Charlotte Hornets  
1   Charlotte Hornets  
2   Charlotte Hornets  
3   Charlotte Hornets  
4   Charlotte Hornets  
5   Charlotte Hornets  
6   Charlotte Hornets  
7   Charlotte Hornets  
8   Charlotte Hornets  
9   Charlotte Hornets  
10  Charlotte Hornets  
11  Charlotte Hornets

In [27]:
team_page_list[18]['team_name']

'New Orleans Pelicans'

In [28]:
pelican_rosters=[]
year='2020.html'

res = requests.get('https://www.basketball-reference.com/teams/NOP/2020.html')

soup = BeautifulSoup(res.content, 'lxml')
table = soup.find('table',{'id':'roster'}).find('tbody')
for row in table.find_all('tr')[0:]:        
    player={}        
    player['name']=row.find('td').find('a').text
    player['pos'] =row.find_all('td')[1].text
    player['team']=team_page_list[18]['team_name']
    player['slug']=row.find('td').find('a').attrs['href']
    player['exp'] =row.find_all('td')[6].text
    pelican_rosters.append(player)
    time.sleep(3)


pelican_players_df = pd.DataFrame(pelican_rosters)

In [29]:
pelican_players_df

exp                      name  pos                       slug  \
0    8             E'Twaun Moore  G-F  /players/m/mooreet01.html   
1    3            Brandon Ingram    F  /players/i/ingrabr01.html   
2    2                Lonzo Ball    G   /players/b/balllo01.html   
3    2                 Josh Hart    G   /players/h/hartjo01.html   
4   10              Jrue Holiday    G  /players/h/holidjr01.html   
5    1             Frank Jackson    G  /players/j/jacksfr01.html   
6    9            Derrick Favors  F-C  /players/f/favorde01.html   
7    1          Kenrich Williams  G-F  /players/w/willike04.html   
8    4             Jahlil Okafor  C-F  /players/o/okafoja01.html   
9    R           Zion Williamson    F  /players/w/willizi01.html   
10   R              Jaxson Hayes    F  /players/h/hayesja02.html   
11   R  Nickeil Alexander-Walker    G  /players/a/alexani01.html   
12  13               J.J. Redick    G  /players/r/redicjj01.html   
13   R              Nicolò Melli    F  /players/m/mellini01.html   

                    team  
0   New Orleans Pelicans  
1   New Orleans Pelicans  
2   New Orleans Pelicans  
3   New Orleans Pelicans  
4   New Orleans Pelicans  
5   New Orleans Pelicans  
6   New Orleans Pelicans  
7   New Orleans Pelicans  
8   New Orleans Pelicans  
9   New Orleans Pelicans  
10  New Orleans Pelicans  
11  New Orleans Pelicans  
12  New Orleans Pelicans  
13  New Orleans Pelicans

In [32]:
total_player_df=pd.concat([players_df,nets_players_df,hornets_players_df,pelican_players_df])

In [36]:
total_player_df.reset_index(inplace=True)

In [38]:
total_player_df.to_csv('total_player_df',index=False)

In [2]:
total_player_df = pd.read_csv('total_player_df')

In [3]:
total_player_df.drop('index',axis=1, inplace=True)

In [4]:
total_player_df

exp                      name  pos                       slug  \
0     8          Chandler Parsons    F  /players/p/parsoch01.html   
1     9               Evan Turner    G  /players/t/turneev01.html   
2     6              Allen Crabbe  G-F  /players/c/crabbal01.html   
3     3           DeAndre' Bembry    F  /players/b/bembrde01.html   
4     3              Damian Jones  F-C  /players/j/jonesda03.html   
5     2              John Collins    F  /players/c/collijo01.html   
6     1                Trae Young    G  /players/y/youngtr01.html   
7     1             Kevin Huerter    G  /players/h/huertke01.html   
8     6                  Alex Len    C    /players/l/lenal01.html   
9     R           De'Andre Hunter    F  /players/h/huntede01.html   
10    R               Cam Reddish    F  /players/r/reddica01.html   
11    R            Bruno Fernando    F  /players/f/fernabr01.html   
12    5             Jabari Parker    F  /players/p/parkeja01.html   
13    3              Jaylen Brown  F-G  /players/b/brownja02.html   
14    2              Jayson Tatum    F  /players/t/tatumja01.html   
15    9            Gordon Hayward  F-G  /players/h/haywago01.html   
16    2              Semi Ojeleye    F  /players/o/ojelese01.html   
17    1           Robert Williams    F  /players/w/williro04.html   
18    5              Marcus Smart    G  /players/s/smartma01.html   
19    R            Romeo Langford    G  /players/l/langfro01.html   
20    R            Grant Williams    F  /players/w/willigr01.html   
21    8              Kemba Walker    G  /players/w/walkeke02.html   
22    R            Carsen Edwards    G  /players/e/edwarca01.html   
23    2              Daniel Theis    C  /players/t/theisda01.html   
24    8               Enes Kanter    C  /players/k/kanteen01.html   
25    1            Brad Wanamaker    G  /players/w/wanambr01.html   
26    2          Denzel Valentine    G  /players/v/valende01.html   
27    3                 Kris Dunn    G   /players/d/dunnkr01.html   
28    2           Lauri Markkanen  F-C  /players/m/markkla01.html   
29    4         Cristiano Felício  F-C  /players/f/feliccr01.html   
..   ..                       ...  ...                        ...   
397   9            Garrett Temple    G  /players/t/templga01.html   
398   1               Theo Pinson    G  /players/p/pinsoth01.html   
399   3               David Nwaba    G  /players/n/nwabada01.html   
400   3            Henry Ellenson    F  /players/e/ellenhe01.html   
401   7    Michael Kidd-Gilchrist    F  /players/k/kiddgmi01.html   
402   6               Cody Zeller  C-F  /players/z/zelleco01.html   
403  11             Nicolas Batum  F-G  /players/b/batumni01.html   
404   8           Bismack Biyombo  C-F  /players/b/biyombi01.html   
405  14           Marvin Williams    F  /players/w/willima02.html   
406   3         Willy Hernangómez    C  /players/h/hernawi01.html   
407   2                Malik Monk    G   /players/m/monkma01.html   
408   2              Dwayne Bacon    G  /players/b/bacondw01.html   
409   1             Miles Bridges  G-F  /players/b/bridgmi02.html   
410   1           Devonte' Graham    G  /players/g/grahade01.html   
411   R             PJ Washington    F  /players/w/washipj01.html   
412   4              Terry Rozier    G  /players/r/roziete01.html   
413   8             E'Twaun Moore  G-F  /players/m/mooreet01.html   
414   3            Brandon Ingram    F  /players/i/ingrabr01.html   
415   2                Lonzo Ball    G   /players/b/balllo01.html   
416   2                 Josh Hart    G   /players/h/hartjo01.html   
417  10              Jrue Holiday    G  /players/h/holidjr01.html   
418   1             Frank Jackson    G  /players/j/jacksfr01.html   
419   9            Derrick Favors  F-C  /players/f/favorde01.html   
420   1          Kenrich Williams  G-F  /players/w/willike04.html   
421   4             Jahlil Okafor  C-F  /players/o/okafoja01.html   
422   R           Zion Williamson    F  /players/w/willizi01.html   
42

##### Removing all the rookies

In [17]:
player_df_no_rooks = total_player_df[total_player_df['exp']!= 'R']

In [18]:
player_df_no_rooks.to_csv('player_df_no_rooks.csv',index=False)

In [4]:
player_df_no_rooks = pd.read_csv('player_df_no_rooks.csv')

In [18]:
player_df_no_rooks.shape

(371, 5)

In [5]:
player_df_no_rooks.head()

exp              name  pos                       slug           team
0    8  Chandler Parsons    F  /players/p/parsoch01.html  Atlanta Hawks
1    9       Evan Turner    G  /players/t/turneev01.html  Atlanta Hawks
2    6      Allen Crabbe  G-F  /players/c/crabbal01.html  Atlanta Hawks
3    3   DeAndre' Bembry    F  /players/b/bembrde01.html  Atlanta Hawks
4    3      Damian Jones  F-C  /players/j/jonesda03.html  Atlanta Hawks

In [74]:
player_df_no_rooks.iloc[158]

exp                             7
name               Meyers Leonard
pos                           C-F
slug    /players/l/leoname01.html
team                   Miami Heat
Name: 158, dtype: object

In [7]:
for i in range(len(player_df_no_rooks)):
    print(player_df_no_rooks.iloc[i])
    print("\n\n\n")

exp                             8
name             Chandler Parsons
pos                             F
slug    /players/p/parsoch01.html
team                Atlanta Hawks
Name: 0, dtype: object




exp                             9
name                  Evan Turner
pos                             G
slug    /players/t/turneev01.html
team                Atlanta Hawks
Name: 1, dtype: object




exp                             6
name                 Allen Crabbe
pos                           G-F
slug    /players/c/crabbal01.html
team                Atlanta Hawks
Name: 2, dtype: object




exp                             3
name              DeAndre' Bembry
pos                             F
slug    /players/b/bembrde01.html
team                Atlanta Hawks
Name: 3, dtype: object




exp                             3
name                 Damian Jones
pos                           F-C
slug    /players/j/jonesda03.html
team                Atlanta Hawks
Name: 4, dtype: object




exp            

Name: 246, dtype: object




exp                             5
name                Tyler Johnson
pos                             G
slug    /players/j/johnsty01.html
team                 Phoenix Suns
Name: 247, dtype: object




exp                             3
name                  Dario Šarić
pos                             F
slug    /players/s/saricda01.html
team                 Phoenix Suns
Name: 248, dtype: object




exp                             1
name                Deandre Ayton
pos                             C
slug    /players/a/aytonde01.html
team                 Phoenix Suns
Name: 249, dtype: object




exp                             1
name                Mikal Bridges
pos                             G
slug    /players/b/bridgmi01.html
team                 Phoenix Suns
Name: 250, dtype: object




exp                             7
name                  Aron Baynes
pos                             C
slug    /players/b/baynear01.html
team                 Phoenix Suns
Name:

In [7]:
url = 'https://www.basketball-reference.com'

##### Scraping the pages of each player. 
##### The two sets of ranges are to scrape every player and to scrape certain sets that were troublesome

In [75]:
stats_of_players=[]
career_stats_of_players =[]


for i in range(158,159): #use this one for testing 
#for i in range(len(player_df_no_rooks)):
    try:
        res = requests.get(url+player_df_no_rooks.iloc[i]['slug'])
        soup = BeautifulSoup(res.content, 'lxml')
        table = soup.find('table',{'id':'per_game'}).find('tbody')


    ###last 3 or so seasons stats
    ## the if statement is to check how many years the player has been in the league
    ## if the player was in the league for over 3 years, then the last 3 years of data are scraped
    ## if the player was for 2 years only 2 years are scraped
    ## if 1 year, then only 1 year is scaped
        if len(table.find_all('tr')) >=3:
            for row in table.find_all('tr',{'class':'full_table'})[-3:]:  #class to catch traded players       
                player={}        
                player['name']=player_df_no_rooks.iloc[i]['name']
                player['season']=row.find('th').find('a').text
                player['age'] = row.find_all('td')[0].text
                #to catch for traded players
                try:
                    player['team'] = row.find_all('td')[1].find('a').text
                except:
                    player['team'] = 'traded'
                player['position'] = row.find_all('td')[3].text
                player['games'] = row.find_all('td')[4].text
                player['games_start'] = row.find_all('td')[5].text
                player['minutes_played'] = row.find_all('td')[6].text
                player['field_goals'] = row.find_all('td')[7].text
                player['field_goals_attempted'] = row.find_all('td')[8].text
                player['field_goal_percentage'] = row.find_all('td')[9].text
                player['3_pointers'] = row.find_all('td')[10].text
                player['3_pointers_attempted'] = row.find_all('td')[11].text
                player['3pt_percentage'] = row.find_all('td')[12].text
                player['2_pointers'] = row.find_all('td')[13].text
                player['2_pointers_attempted'] = row.find_all('td')[14].text
                player['2pt_percentage']= row.find_all('td')[15].text
                player['effective_field_goal_percentage'] = row.find_all('td')[16].text
                player['free_throws'] = row.find_all('td')[17].text
                player['free_throws_attempted'] = row.find_all('td')[18].text
                player['free_throw_percentage'] = row.find_all('td')[19].text
                player['offensive_rebounds'] = row.find_all('td')[20].text
                player['defensive_rebounds'] = row.find_all('td')[21].text
                player['total_rebounds'] = row.find_all('td')[22].text
                player['assists'] = row.find_all('td')[23].text
                player['steals'] = row.find_all('td')[24].text
                player['blocks'] = row.find_all('td')[25].text
                player['turnovers'] = row.find_all('td')[26].text
                player['personal_fouls'] = row.find_all('td')[27].text
                player['points'] = row.find_all('td')[28].text
                #this is for players that have not attempted a 3
                if ((row.find_all('td')[10].text == '0.0') &
                    (row.find_all('td')[11].text=='0.0')):
                    player['3pt_percentage'] = '0'
                stats_of_players.append(player)
        elif len(table.find_all('tr',{'class':'full_table'}))==2:
            for row in table.find_all('tr')[-2:]:        
                player={}        
                player['name']=player_df_no_rooks.iloc[i]['name']
                player['season']=row.find('th').find('a').text
                player['age'] = row.find_all('td')[0].text
                try:
                    player['team'] = row.find_all('td')[1].find('a').text
                except:
                    player['team'] = 'traded'
                player['position'] = row.find_all('td')[3].text
                player['games'] = row.find_all('td')[4].text
                player['games_start'] = row.find_all('td')[5].text
                player['minutes_played'] = row.find_all('td')[6].text
                player['field_goals'] = row.find_all('td')[7].text
                player['field_goals_attempted'] = row.find_all('td')[8].text
                player['field_goal_percentage'] = row.find_all('td')[9].text
                player['3_pointers'] = row.find_all('td')[10].text
                player['3_pointers_attempted'] = row.find_all('td')[11].text
                player['3pt_percentage'] = row.find_all('td')[12].text
                player['2_pointers'] = row.find_all('td')[13].text
                player['2_pointers_attempted'] = row.find_all('td')[14].text
                player['2pt_percentage']= row.find_all('td')[15].text
                player['effective_field_goal_percentage'] = row.find_all('td')[16].text
                player['free_throws'] = row.find_all('td')[17].text
                player['free_throws_attempted'] = row.find_all('td')[18].text
                player['free_throw_percentage'] = row.find_all('td')[19].text
                player['offensive_rebounds'] = row.find_all('td')[20].text
                player['defensive_rebounds'] = row.find_all('td')[21].text
                player['total_rebounds'] = row.find_all('td')[22].text
                player['assists'] = row.find_all('td')[23].text
                player['steals'] = row.find_all('td')[24].text
                player['blocks'] = row.find_all('td')[25].text
                player['turnovers'] = row.find_all('td')[26].text
                player['personal_fouls'] = row.find_all('td')[27].text
                player['points'] = row.find_all('td')[28].text
                #this is for players that have not attempted a 3
                if ((row.find_all('td')[10].text == '0.0') &
                    (row.find_all('td')[11].text=='0.0')):
                    player['3pt_percentage'] = '0'

                stats_of_players.append(player)
        elif len(table.find_all('tr',{'class':'full_table'}))==1:
            for row in table.find_all('tr')[-1:]:        
                player={}        
                player['name']=player_df_no_rooks.iloc[i]['name']
                player['season']=row.find('th').find('a').text
                player['age'] = row.find_all('td')[0].text
                try:
                    player['team'] = row.find_all('td')[1].find('a').text
                except:
                    player['team'] = 'traded'
                player['position'] = row.find_all('td')[3].text
                player['games'] = row.find_all('td')[4].text
                player['games_start'] = row.find_all('td')[5].text
                player['minutes_played'] = row.find_all('td')[6].text
                player['field_goals'] = row.find_all('td')[7].text
                player['field_goals_attempted'] = row.find_all('td')[8].text
                player['field_goal_percentage'] = row.find_all('td')[9].text
                player['3_pointers'] = row.find_all('td')[10].text
                player['3_pointers_attempted'] = row.find_all('td')[11].text
                player['3pt_percentage'] = row.find_all('td')[12].text
                player['2_pointers'] = row.find_all('td')[13].text
                player['2_pointers_attempted'] = row.find_all('td')[14].text
                player['2pt_percentage']= row.find_all('td')[15].text
                player['effective_field_goal_percentage'] = row.find_all('td')[16].text
                player['free_throws'] = row.find_all('td')[17].text
                player['free_throws_attempted'] = row.find_all('td')[18].text
                player['free_throw_percentage'] = row.find_all('td')[19].text
                player['offensive_rebounds'] = row.find_all('td')[20].text
                player['defensive_rebounds'] = row.find_all('td')[21].text
                player['total_rebounds'] = row.find_all('td')[22].text
                player['assists'] = row.find_all('td')[23].text
                player['steals'] = row.find_all('td')[24].text
                player['blocks'] = row.find_all('td')[25].text
                player['turnovers'] = row.find_all('td')[26].text
                player['personal_fouls'] = row.find_all('td')[27].text
                player['points'] = row.find_all('td')[28].text
                #this is for players that have not attempted a 3
                if ((row.find_all('td')[10].text == '0.0') &
                    (row.find_all('td')[11].text=='0.0')):
                    player['3pt_percentage'] = '0'
                stats_of_players.append(player)


    ### Career stats gathering
        table = soup.find('table',{'id':'per_game'}).find('tfoot')
        player={}
        player['name']=player_df_no_rooks.iloc[i]['name']    
        player['games'] = table.find_all('td')[4].text
        player['games_start'] = table.find_all('td')[5].text
        player['minutes_played'] = table.find_all('td')[6].text
        player['field_goals'] = table.find_all('td')[7].text
        player['field_goals_attempted'] = table.find_all('td')[8].text
        player['field_goal_percentage'] = table.find_all('td')[9].text
        player['3_pointers'] = table.find_all('td')[10].text
        player['3_pointers_attempted'] = table.find_all('td')[11].text
        player['3pt_percentage'] = table.find_all('td')[12].text
        player['2_pointers'] = table.find_all('td')[13].text
        player['2_pointers_attempted'] = table.find_all('td')[14].text
        player['2pt_percentage']= table.find_all('td')[15].text
        player['effective_field_goal_percentage'] = table.find_all('td')[16].text
        player['free_throws'] = table.find_all('td')[17].text
        player['free_throws_attempted'] = table.find_all('td')[18].text
        player['free_throw_percentage'] = table.find_all('td')[19].text
        player['offensive_rebounds'] = table.find_all('td')[20].text
        player['defensive_rebounds'] = table.find_all('td')[21].text
        player['total_rebounds'] = table.find_all('td')[22].text
        player['assists'] = table.find_all('td')[23].text
        player['steals'] = table.find_all('td')[24].text
        player['blocks'] = table.find_all('td')[25].text
        player['turnovers'] = table.find_all('td')[26].text
        player['personal_fouls'] = table.find_all('td')[27].text
        player['points'] = table.find_all('td')[28].text
        #this is for players that have not attempted a 3
        if ((row.find_all('td')[10].text == '0.0') &
            (row.find_all('td')[11].text=='0.0')):
            player['3pt_percentage'] = '0'
        career_stats_of_players.append(player)
        #time.sleep(15)

    except:
            print(f'blame the guy at {i}')

### <u>Challenges with scrape</u>
I wanted to get the last 3 seasons of players however there are plenty of players with less than 3 years of experience. The first if statement is to catch if the players had 3+ years experience, 2 years, and 1 year. 

The try except in the loop is due to the code erroring out when looking at the team value of a player. If a player was traded mid season, the code would error. The try except catches and deals with this error.

The if statement in the loop is to put a value in "3pt percentage" if the player did not attempt any 3's. Without this, the value of the cell would be blank and not show up in isnull().

The outside try/except is to catch if there an error in the attempt at scraping. These errors could be due to trying to access the source page too much. The except shows which indicies errors occured.

During this process error occured between indicies 38-158 out of a list of 370 players. 

The script was run again for the indicies 38-158.

It was then run again to catch one player that slipped through the cracks

In [13]:
# remaining_stats_of_players_df = pd.DataFrame(stats_of_players)
# remaining_career_stats_of_players_df = pd.DataFrame(career_stats_of_players)

In [76]:
# last guy
last_guy_stats_of_players_df = pd.DataFrame(stats_of_players)
last_guy_career_stats_of_players_df = pd.DataFrame(career_stats_of_players)

In [50]:
#38-158 were not scraped initially

In [51]:
stats_of_players_df.columns

Index(['2_pointers', '2_pointers_attempted', '2pt_percentage', '3_pointers',
       '3_pointers_attempted', '3pt_percentage', 'age', 'assists', 'blocks',
       'defensive_rebounds', 'effective_field_goal_percentage',
       'field_goal_percentage', 'field_goals', 'field_goals_attempted',
       'free_throw_percentage', 'free_throws', 'free_throws_attempted',
       'games', 'games_start', 'minutes_played', 'name', 'offensive_rebounds',
       'personal_fouls', 'points', 'position', 'season', 'steals', 'team',
       'total_rebounds', 'turnovers'],
      dtype='object')

In [16]:
remaining_stats_of_players_df[['name',
                     'age','team', 'position',
                     'games','games_start',
                     'minutes_played',
                     'field_goals','field_goals_attempted','field_goal_percentage',
                     '3_pointers','3_pointers_attempted',
                     '3pt_percentage',
                     '2_pointers','2_pointers_attempted',
                     '2pt_percentage','effective_field_goal_percentage',
                     'free_throws','free_throws_attempted','free_throw_percentage',
                     'offensive_rebounds','defensive_rebounds',
                     'total_rebounds','assists',
                     'steals','blocks','turnovers','personal_fouls',
                     
                     'points']]

name age    team position games games_start  \
0     Matthew Dellavedova  26     MIL       PG    76          54   
1     Matthew Dellavedova  27     MIL       PG    38           3   
2     Matthew Dellavedova  28  traded       PG    48           0   
3              Ante Žižić  21     CLE        C    32           2   
4              Ante Žižić  22     CLE        C    59          25   
5              Cedi Osman  22     CLE       SF    61          12   
6              Cedi Osman  23     CLE       SF    76          75   
7           Collin Sexton  20     CLE       PG    82          72   
8              Kevin Love  28     CLE       PF    60          60   
9              Kevin Love  29     CLE        C    59          59   
10             Kevin Love  30     CLE       PF    22          21   
11        Larry Nance Jr.  24     LAL       PF    63           7   
12        Larry Nance Jr.  25  traded        C    66          27   
13        Larry Nance Jr.  26     CLE        C    67          30   
14          Dwight Powell  25     DAL        C    77           3   
15          Dwight Powell  26     DAL        C    79          25   
16          Dwight Powell  27     DAL        C    77          22   
17           Courtney Lee  31     NYK       SG    77          74   
18           Courtney Lee  32     NYK       SG    76          69   
19           Courtney Lee  33  traded       SG    34           6   
20         Justin Jackson  22     SAC       SF    68          41   
21         Justin Jackson  23  traded       SF    81          14   
22       Tim Hardaway Jr.  24     ATL       SG    79          30   
23       Tim Hardaway Jr.  25     NYK       SG    57          54   
24       Tim Hardaway Jr.  26  traded       SG    65          63   
25            Luka Dončić  19     DAL       SG    72          72   
26         Ryan Broekhoff  28     DAL       SG    42           0   
27          Jalen Brunson  22     DAL       PG    73          38   
28   Kostas Antetokounmpo  21     DAL       PF     2           0   
29            Daryl Macon  23     DAL       SG     8           0   
..                    ...  ..     ...      ...   ...         ...   
270          Josh Jackson  21     PHO       SG    79          29   
271        Andre Iguodala  33     GSW       SF    76           0   
272        Andre Iguodala  34     GSW       SF    64           7   
273        Andre Iguodala  35     GSW       SF    68          13   
274         Dillon Brooks  22     MEM       SF    82          74   
275         Dillon Brooks  23     MEM       SF    18           0   
276             Ivan Rabb  20     MEM       PF    36           5   
277             Ivan Rabb  21     MEM       PF    49          13   
278     Jaren Jackson Jr.  19     MEM       PF    58          56   
279         Grayson Allen  23     UTA       SG    38           2   
280         Kyle Anderson  23     SAS       SF    72          14   
281         Kyle Anderson  24     SAS       SF    74          67   
282         Kyle Anderson  25     MEM       SF    43          40   
283         Dwight Howard  31     ATL        C    74          74   
284         Dwight Howard  32     CHO        C    81          81   
285         Dwight Howard  33     WAS        C     9           9   
286         Yuta Watanabe  24     MEM       SF    15           0   
287     De'Anthony Melton  20     PHO       PG    50          31   
288         Bruno Caboclo  21     TOR       SF     9           0   
289         Bruno Caboclo  22  traded       SF    12           0   
290         Bruno Caboclo  23     MEM       SF    34          19   
291            Tyus Jones  20     MIN       PG    60           0   
292            Tyus Jones  21     MIN       PG    82          11   
293            Tyus Jones  22     MIN       PG    68          23   
294     Jonas Valančiūnas  24     TOR        C    80          80   
295     Jonas Valančiūnas  25     TOR        C    77          77   
296     Jonas Valančiūnas  26  traded        C    49          27   
297      

In [14]:
stats_of_players_df[['name',
                     'age','team', 'position',
                     'games','games_start',
                     'minutes_played',
                     'field_goals','field_goals_attempted','field_goal_percentage',
                     '3_pointers','3_pointers_attempted',
                     '3pt_percentage',
                     '2_pointers','2_pointers_attempted',
                     '2pt_percentage','effective_field_goal_percentage',
                     'free_throws','free_throws_attempted','free_throw_percentage',
                     'offensive_rebounds','defensive_rebounds',
                     'total_rebounds','assists',
                     'steals','blocks','turnovers','personal_fouls',
                     
                     'points']]

name  age    team position  games  games_start  \
0    Chandler Parsons   28     MEM       SF     34           34   
1    Chandler Parsons   29     MEM       SF     36            8   
2    Chandler Parsons   30     MEM       SF     25            3   
3         Evan Turner   28     POR       SF     65           12   
4         Evan Turner   29     POR       SF     79           40   
5         Evan Turner   30     POR       PG     73            2   
6        Allen Crabbe   24     POR       SG     79            7   
7        Allen Crabbe   25     BRK       SG     75           68   
8        Allen Crabbe   26     BRK       SG     43           20   
9     DeAndre' Bembry   22     ATL       SF     38            1   
10    DeAndre' Bembry   23     ATL       SF     26            3   
11    DeAndre' Bembry   24     ATL       SG     82           15   
12       Damian Jones   21     GSW        C     10            0   
13       Damian Jones   22     GSW        C     15            0   
14       Damian Jones   23     GSW        C     24           22   
15       John Collins   20     ATL       PF     74           26   
16       John Collins   21     ATL       PF     61           59   
17         Trae Young   20     ATL       PG     81           81   
18      Kevin Huerter   20     ATL       SG     75           59   
19           Alex Len   23     PHO        C     77           34   
20           Alex Len   24     PHO        C     69           13   
21           Alex Len   25     ATL        C     77           31   
22      Jabari Parker   21     MIL       PF     51           50   
23      Jabari Parker   22     MIL       PF     31            3   
24      Jabari Parker   23  traded       PF     64           17   
25       Jaylen Brown   20     BOS       SF     78           20   
26       Jaylen Brown   21     BOS       SG     70           70   
27       Jaylen Brown   22     BOS       SG     74           25   
28       Jayson Tatum   19     BOS       SF     80           80   
29       Jayson Tatum   20     BOS       SF     79           79   
..                ...  ...     ...      ...    ...          ...   
578      Dwayne Bacon   23     CHO       SG     43           13   
579     Miles Bridges   20     CHO       SF     80           25   
580   Devonte' Graham   23     CHO       PG     46            3   
581      Terry Rozier   22     BOS       PG     74            0   
582      Terry Rozier   23     BOS       PG     80           16   
583      Terry Rozier   24     BOS       PG     79           14   
584     E'Twaun Moore   27     NOP       SG     73           22   
585     E'Twaun Moore   28     NOP       SG     82           80   
586     E'Twaun Moore   29     NOP       SG     53           36   
587    Brandon Ingram   19     LAL       SF     79           40   
588    Brandon Ingram   20     LAL       SF     59           59   
589    Brandon Ingram   21     LAL       SF     52           52   
590        Lonzo Ball   20     LAL       PG     52           50   
591        Lonzo Ball   21     LAL       PG     47           45   
592         Josh Hart   22     LAL       SG     63           23   
593         Josh Hart   23     LAL       SG     67           22   
594      Jrue Holiday   26     NOP       PG     67           61   
595      Jrue Holiday   27     NOP       SG     81           81   
596      Jrue Holiday   28     NOP       SG     67           67   
597     Frank Jackson   20     NOP       PG     61           16   
598    Derrick Favors   25     UTA       PF     50           39   
599    Derrick Favors   26     UTA        C     77           77   
600    Derrick Favors   27     UTA        C     76           70   
601  Kenrich Williams   24     NOP       SF     46           29   
602     Jahlil Okafor   21     PHI        C     50           33   
603     Jahlil Okafor   22  traded        C     28            0   
604     Jahlil Okafor   23     NOP        C     59           24   
605       J.J. Redick   32     LAC       SG     78           78 

In [55]:
stats_of_players_df.to_csv("partial_stats_of_players_df.csv")

In [8]:
stats_of_players_df=pd.read_csv("partial_stats_of_players_df.csv")

In [15]:
career_stats_of_players_df[['name',

                     'games','games_start',
                     'minutes_played',
                     'field_goals','field_goals_attempted','field_goal_percentage',
                     '3_pointers','3_pointers_attempted',
                     '3pt_percentage',
                     '2_pointers','2_pointers_attempted',
                     '2pt_percentage','effective_field_goal_percentage',
                     'free_throws','free_throws_attempted','free_throw_percentage',
                     'offensive_rebounds','defensive_rebounds',
                     'total_rebounds','assists',
                     'steals','blocks','turnovers','personal_fouls',
                     
                     'points']]

name  games  games_start  minutes_played  field_goals  \
0          Chandler Parsons    435          369            30.3          4.9   
1               Evan Turner    686          295            27.2          4.0   
2              Allen Crabbe    344          112            24.6          3.4   
3           DeAndre' Bembry    146           19            18.9          2.6   
4              Damian Jones     49           22            11.9          1.5   
5              John Collins    135           85            26.8          5.8   
6                Trae Young     81           81            30.9          6.5   
7             Kevin Huerter     75           59            27.3          3.7   
8                  Alex Len    412          171            19.9          3.0   
9             Jabari Parker    247          167            29.7          6.1   
10             Jaylen Brown    222          115            24.4          4.2   
11             Jayson Tatum    159          159            30.8          5.4   
12           Gordon Hayward    589          427            30.6          5.1   
13             Semi Ojeleye    129            3            13.5          1.0   
14          Robert Williams     32            2             8.8          1.1   
15             Marcus Smart    341          143            28.4          3.1   
16             Kemba Walker    605          560            34.1          6.9   
17             Daniel Theis    129            5            14.3          2.1   
18              Enes Kanter    583          216            22.2          4.9   
19           Brad Wanamaker     36            0             9.5          1.4   
20         Denzel Valentine    134           37            22.9          3.0   
21                Kris Dunn    176           94            24.1          3.5   
22          Lauri Markkanen    120          119            30.9          6.0   
23        Cristiano Felício    212           20            14.5          1.8   
24              Otto Porter    399          286            26.9          4.3   
25       Chandler Hutchison     44           14            20.3          2.2   
26       Wendell Carter Jr.     44           44            25.2          4.1   
27              Zach LaVine    293          206            29.9          5.8   
28         Antonio Blakeney     76            3            15.0          2.9   
29         Tomáš Satoranský    210           87            21.6          2.5   
..                      ...    ...          ...             ...          ...   
220       Spencer Dinwiddie    253           81            24.4          3.5   
221            Kevin Durant    849          849            36.9          9.2   
222          DeAndre Jordan    819          707            28.2          3.8   
223            Kyrie Irving    508          508            33.8          8.2   
224         Wilson Chandler    641          480            30.5          5.0   
225          Garrett Temple    514          212            20.3          2.1   
226             Theo Pinson     18            0            11.7          1.4   
227             David Nwaba    141           37            21.5          2.7   
228          Henry Ellenson     76            2             9.7          1.5   
229  Michael Kidd-Gilchrist    421          356            25.4          3.4   
230             Cody Zeller    361          213            22.9          3.1   
231           Nicolas Batum    767          711            31.6          4.2   
232         Bismack Biyombo    583          242            20.3          1.9   
233         Marvin Williams   1014          827            28.6          3.8   
234       Willy Hernangómez    178           26            14.8          2.8   
235              Malik Monk    136            0            15.5          2.8   
236            Dwayne Bacon     96           19            15.3          2.0   
237           Miles Bridges     80           25            21.2          3.0   
238         Devonte' Gr

In [17]:
remaining_career_stats_of_players_df.shape

(120, 26)

In [10]:
career_stats_of_players_df.shape

(250, 27)

In [56]:
career_stats_of_players_df.to_csv('partial_career_stats_of_players_df.csv')

In [9]:
career_stats_of_players_df=pd.read_csv('partial_career_stats_of_players_df.csv')

In [58]:
career_stats_of_players_df.shape

(250, 27)

In [19]:
career_stats_of_players_df.head()

Unnamed: 0  2_pointers  2_pointers_attempted  2pt_percentage  3_pointers  \
0           0         3.3                   6.4           0.522         1.6   
1           1         3.6                   7.8           0.458         0.4   
2           2         1.7                   3.6           0.472         1.7   
3           3         2.1                   4.3           0.504         0.4   
4           4         1.5                   2.3           0.643         0.0   

   3_pointers_attempted  3pt_percentage  assists  blocks  defensive_rebounds  \
0                   4.2           0.373      2.8     0.3                 3.7   
1                   1.3           0.296      3.5     0.3                 4.2   
2                   4.3           0.393      1.2     0.3                 2.6   
3                   1.5           0.281      1.9     0.4                 2.8   
4                   0.0           0.000      0.6     0.7                 1.4   

   effective_field_goal_percentage  field_goal_percentage  field_goals  \
0                            0.537                  0.462          4.9   
1                            0.456                  0.435          4.0   
2                            0.536                  0.429          3.4   
3                            0.482                  0.445          2.6   
4                            0.643                  0.643          1.5   

   field_goals_attempted  free_throw_percentage  free_throws  \
0                   10.6                  0.713          1.4   
1                    9.1                  0.781          1.5   
2                    7.9                  0.835          1.1   
3                    5.8                  0.605          0.8   
4                    2.3                  0.577          0.6   

   free_throws_attempted  games  games_start  minutes_played  \
0                    2.0    435          369            30.3   
1                    2.0    686          295            27.2   
2                    1.3    344          112            24.6   
3                    1.3    146           19            18.9   
4                    1.1     49           22            11.9   

               name  offensive_rebounds  personal_fouls  points  steals  \
0  Chandler Parsons                 0.8             2.0    12.8     0.9   
1       Evan Turner                 0.6             2.0     9.9     0.8   
2      Allen Crabbe                 0.3             2.1     9.5     0.6   
3   DeAndre' Bembry                 0.5             1.7     6.3     0.9   
4      Damian Jones                 0.9             1.9     3.6     0.3   

   total_rebounds  turnovers  
0             4.5        1.5  
1             4.7        1.8  
2             2.9        0.8  
3             3.3        1.4  
4             2.3        0.5

In [20]:
career_stats_of_players_df.drop(columns='Unnamed: 0',inplace=True)

In [25]:
total_stats_of_players_df=pd.concat([stats_of_players_df,remaining_stats_of_players_df])

In [22]:
total_career_stats_of_players_df=pd.concat([career_stats_of_players_df,remaining_career_stats_of_players_df])

In [30]:
total_career_stats_of_players_df.head()

2_pointers 2_pointers_attempted 2pt_percentage 3_pointers  \
0        3.3                  6.4          0.522        1.6   
1        3.6                  7.8          0.458        0.4   
2        1.7                  3.6          0.472        1.7   
3        2.1                  4.3          0.504        0.4   
4        1.5                  2.3          0.643          0   

  3_pointers_attempted 3pt_percentage assists blocks defensive_rebounds  \
0                  4.2          0.373     2.8    0.3                3.7   
1                  1.3          0.296     3.5    0.3                4.2   
2                  4.3          0.393     1.2    0.3                2.6   
3                  1.5          0.281     1.9    0.4                2.8   
4                    0              0     0.6    0.7                1.4   

  effective_field_goal_percentage field_goal_percentage field_goals  \
0                           0.537                 0.462         4.9   
1                           0.456                 0.435           4   
2                           0.536                 0.429         3.4   
3                           0.482                 0.445         2.6   
4                           0.643                 0.643         1.5   

  field_goals_attempted free_throw_percentage free_throws  \
0                  10.6                 0.713         1.4   
1                   9.1                 0.781         1.5   
2                   7.9                 0.835         1.1   
3                   5.8                 0.605         0.8   
4                   2.3                 0.577         0.6   

  free_throws_attempted games games_start minutes_played              name  \
0                     2   435         369           30.3  Chandler Parsons   
1                     2   686         295           27.2       Evan Turner   
2                   1.3   344         112           24.6      Allen Crabbe   
3                   1.3   146          19           18.9   DeAndre' Bembry   
4                   1.1    49          22           11.9      Damian Jones   

  offensive_rebounds personal_fouls points steals total_rebounds turnovers  
0                0.8              2   12.8    0.9            4.5       1.5  
1                0.6              2    9.9    0.8            4.7       1.8  
2                0.3            2.1    9.5    0.6            2.9       0.8  
3                0.5            1.7    6.3    0.9            3.3       1.4  
4                0.9            1.9    3.6    0.3            2.3       0.5

In [36]:
total_career_stats_of_players_df.reset_index(inplace=True)

In [44]:
total_career_stats_of_players_df['name']

0              Chandler Parsons
1                   Evan Turner
2                  Allen Crabbe
3               DeAndre' Bembry
4                  Damian Jones
5                  John Collins
6                    Trae Young
7                 Kevin Huerter
8                      Alex Len
9                 Jabari Parker
10                 Jaylen Brown
11                 Jayson Tatum
12               Gordon Hayward
13                 Semi Ojeleye
14              Robert Williams
15                 Marcus Smart
16                 Kemba Walker
17                 Daniel Theis
18                  Enes Kanter
19               Brad Wanamaker
20             Denzel Valentine
21                    Kris Dunn
22              Lauri Markkanen
23            Cristiano Felício
24                  Otto Porter
25           Chandler Hutchison
26           Wendell Carter Jr.
27                  Zach LaVine
28             Antonio Blakeney
29             Tomáš Satoranský
                 ...           
340     

##### The follwing code was to find which player slipped through the scrape

In [72]:
#used this function to figure out the name of the person who was accidentally cut off
for name in player_df_no_rooks['name']:
    if name not in list(total_career_stats_of_players_df['name']):
        print(name)


Meyers Leonard


In [39]:
player_df_no_rooks['name'][0]==total_career_stats_of_players_df['name'][0]

True

In [32]:
total_career_stats_of_players_df.to_csv("total_career_stats_of_players_df.csv",index=False)
total_stats_of_players_df.to_csv("total_stats_of_players_df.csv",index=False)

In [106]:
total_stats_of_players_df=pd.read_csv("total_stats_of_players_df.csv")

(908, 30)

In [55]:
total_career_stats_of_players_df.duplicated().sum()

0

In [58]:
total_stats_of_players_df[total_stats_of_players_df.duplicated()]

2_pointers 2_pointers_attempted 2pt_percentage 3_pointers  \
359        2.7                  6.1           0.44        1.9   
360        3.3                    7          0.467        2.1   
361        2.9                    6          0.478        1.2   

    3_pointers_attempted 3pt_percentage age assists blocks defensive_rebounds  \
359                  5.2          0.371  24     1.6    0.2                3.2   
360                  5.6          0.381  25     1.6    0.2                2.4   
361                  3.3          0.356  26     1.8    0.2                1.8   

    effective_field_goal_percentage field_goal_percentage field_goals  \
359                           0.494                 0.408         4.6   
360                           0.513                 0.429         5.4   
361                           0.498                 0.435         4.1   

    field_goals_attempted free_throw_percentage free_throws  \
359                  11.3                 0.783         1.5   
360                  12.5                  0.86         1.9   
361                   9.3                 0.884         1.9   

    free_throws_attempted games games_start minutes_played         name  \
359                   1.9    59          55             27  Rodney Hood   
360                   2.2    60          23           26.9  Rodney Hood   
361                   2.2    72          49           26.3  Rodney Hood   

    offensive_rebounds personal_fouls points position   season steals    team  \
359                0.3            2.2   12.7       SG  2016-17    0.6     UTA   
360                0.4            1.8   14.7       SG  2017-18    0.8  traded   
361                0.3              2   11.2       SG  2018-19    0.8  traded   

    total_rebounds turnovers  
359            3.4       1.1  
360            2.8       1.2  
361            2.2       0.8

In [104]:
last_guy_career_stats_of_players_df

2_pointers 2_pointers_attempted 2pt_percentage 3_pointers  \
0        1.4                  2.7           .542        0.7   

  3_pointers_attempted 3pt_percentage assists blocks defensive_rebounds  \
0                  1.8           .385     0.9    0.3                3.0   

  effective_field_goal_percentage field_goal_percentage field_goals  \
0                            .556                  .479         2.1   

  field_goals_attempted free_throw_percentage free_throws  \
0                   4.4                  .828         0.6   

  free_throws_attempted games games_start minutes_played            name  \
0                   0.7   393          42           15.5  Meyers Leonard   

  offensive_rebounds personal_fouls points steals total_rebounds turnovers  
0                0.7            2.2    5.6    0.2            3.7       0.7

In [78]:
last_guy_stats_of_players_df

2_pointers 2_pointers_attempted 2pt_percentage 3_pointers  \
0        1.0                  2.2           .436        1.0   
1        1.1                  1.6           .673        0.3   
2        1.3                  2.1           .626        0.8   

  3_pointers_attempted 3pt_percentage age assists blocks defensive_rebounds  \
0                  2.9           .347  24     1.0    0.4                2.8   
1                  0.8           .423  25     0.5    0.0                1.8   
2                  1.8           .450  26     1.2    0.1                3.0   

  effective_field_goal_percentage field_goal_percentage field_goals  \
0                            .484                  .386         2.0   
1                            .660                  .590         1.4   
2                            .649                  .545         2.2   

  field_goals_attempted free_throw_percentage free_throws  \
0                   5.1                  .875         0.5   
1                   2.4                  .818         0.3   
2                   4.0                  .843         0.7   

  free_throws_attempted games games_start minutes_played            name  \
0                   0.5    74          12           16.5  Meyers Leonard   
1                   0.3    33           2            7.7  Meyers Leonard   
2                   0.8    61           2           14.4  Meyers Leonard   

  offensive_rebounds personal_fouls points position   season steals team  \
0                0.4            2.3    5.4       PF  2016-17    0.2  POR   
1                0.4            1.1    3.4        C  2017-18    0.2  POR   
2                0.8            1.7    5.9        C  2018-19    0.2  POR   

  total_rebounds turnovers  
0            3.2       0.5  
1            2.1       0.3  
2            3.8       0.7

In [80]:
### this is to add the last person that was unfortunately cut off due to indexing
total_career_stats_of_players_df=pd.concat([total_career_stats_of_players_df,last_guy_career_stats_of_players_df])

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


In [109]:
total_stats_of_players_df=pd.concat([total_stats_of_players_df,last_guy_stats_of_players_df])

##### Making sure that all players present

In [110]:
player_df_no_rooks.shape

(371, 5)

In [111]:
total_career_stats_of_players_df.shape

(371, 26)

In [112]:
total_stats_of_players_df.shape

(911, 30)

In [92]:
total_stats_of_players_df.columns

Index(['2_pointers', '2_pointers_attempted', '2pt_percentage', '3_pointers',
       '3_pointers_attempted', '3pt_percentage', 'age', 'assists', 'blocks',
       'defensive_rebounds', 'effective_field_goal_percentage',
       'field_goal_percentage', 'field_goals', 'field_goals_attempted',
       'free_throw_percentage', 'free_throws', 'free_throws_attempted',
       'games', 'games_start', 'minutes_played', 'name', 'offensive_rebounds',
       'personal_fouls', 'points', 'position', 'season', 'steals', 'team',
       'total_rebounds', 'turnovers'],
      dtype='object')

In [95]:
total_career_stats_of_players_df=total_career_stats_of_players_df[['name',

                     'games','games_start',
                     'minutes_played',
                     'field_goals','field_goals_attempted','field_goal_percentage',
                     '3_pointers','3_pointers_attempted',
                     '3pt_percentage',
                     '2_pointers','2_pointers_attempted',
                     '2pt_percentage','effective_field_goal_percentage',
                     'free_throws','free_throws_attempted','free_throw_percentage',
                     'offensive_rebounds','defensive_rebounds',
                     'total_rebounds','assists',
                     'steals','blocks','turnovers','personal_fouls',
                     
                     'points']]

In [113]:
total_stats_of_players_df=total_stats_of_players_df[['name',

                     'games','games_start',
                     'minutes_played',
                     'field_goals','field_goals_attempted','field_goal_percentage',
                     '3_pointers','3_pointers_attempted',
                     '3pt_percentage',
                     '2_pointers','2_pointers_attempted',
                     '2pt_percentage','effective_field_goal_percentage',
                     'free_throws','free_throws_attempted','free_throw_percentage',
                     'offensive_rebounds','defensive_rebounds',
                     'total_rebounds','assists',
                     'steals','blocks','turnovers','personal_fouls',
                     
                     'points']]

In [115]:
total_stats_of_players_df.shape

(911, 26)

In [116]:
total_career_stats_of_players_df.shape

(371, 26)

In [117]:
total_career_stats_of_players_df.to_csv("total_career_stats_of_players_df.csv",index=False)


total_stats_of_players_df.to_csv("total_stats_of_players_df.csv",index=False)